feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,593 @@
+		Redirect:      getTcDataFromPccRule(policyDecision, pccRule),
+		Qos:           getQosDataFromPccRule(policyDecision, pccRule),
+		FlowList:      getFlowListFromPccRule(pccRule),
+	}
+	insertChargingDataToPolicyRule(policyDecision, pccRule, policyRule)
+	policyRule.TrackingType = getTrackingTypeProto(len(policyRule.MonitoringKey) != 0, policyRule.RatingGroup != 0)
+
+	if len(policyRule.MonitoringKey) == 0 && policyRule.Redirect == nil &&
+		policyRule.Qos == nil && len(policyRule.FlowList) == 0 && policyRule.RatingGroup == 0 {
+		// must be a static rule if these are not present
+		return nil
+	}
+	return policyRule
+}
+
+func insertChargingDataToPolicyRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+	policyRule *protos.PolicyRule,
+) {
+	if pccRule.RefChgData == nil || len(*pccRule.RefChgData) == 0 || policyDecision.ChgDecs == nil {
+		return
+	}
+	chgData, found := policyDecision.ChgDecs.Get((*pccRule.RefChgData)[0])
+	if !found {
+		return
+	}
+	policyRule.RatingGroup = uint32(*chgData.RatingGroup)
+	policyRule.ServiceIdentifier = &protos.ServiceIdentifier{
+		Value: uint32(*chgData.ServiceId),
+	}
+	policyRule.Offline = getSbiBool(chgData.Offline)
+	policyRule.Online = getSbiBool(chgData.Online)
+}
+
+func getMonKeyFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) []byte {
+	if pccRule.RefUmData != nil && len(*pccRule.RefUmData) > 0 && policyDecision.UmDecs != nil {
+		umData, found := policyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+		if found {
+			return []byte(umData.UmId)
+		}
+	}
+	return []byte{}
+}
+
+func getTcDataFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.RedirectInformation {
+	if pccRule.RefTcData != nil && len(*pccRule.RefTcData) > 0 && policyDecision.TraffContDecs != nil {
+		tcData, found := policyDecision.TraffContDecs.Get((*pccRule.RefTcData)[0])
+		if found {
+			return GetRedirectInfoProto(tcData.RedirectInfo)
+		}
+	}
+	return nil
+}
+
+func getQosDataFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.FlowQos {
+	if pccRule.RefQosData != nil && len(*pccRule.RefQosData) > 0 && policyDecision.QosDecs != nil {
+		qosData, found := policyDecision.QosDecs.Get((*pccRule.RefQosData)[0])
+		if found {
+			return GetFlowQosProto(&qosData)
+		}
+	}
+	return nil
+}
+
+func getFlowListFromPccRule(pccRule *n7_sbi.PccRule) []*protos.FlowDescription {
+	var flowList []*protos.FlowDescription
+	if pccRule.FlowInfos != nil {
+		for _, flowInfo := range *pccRule.FlowInfos {
+			if flowInfo.FlowDescription != nil {
+				flow, err := policydb.GetFlowDescriptionFromFlowString(string(*flowInfo.FlowDescription))
+				if err != nil {
+					glog.Errorf("Could not get flow for description %s : %s", *flowInfo.FlowDescription, err)
+				} else {
+					flowList = append(flowList, flow)
+				}
+			}
+		}
+	}
+	return flowList
+}
```

Comment:
can you un nest this function please?
```
func getFlowListFromPccRule(pccRule *n7_sbi.PccRule) []*protos.FlowDescription {
	var flowList []*protos.FlowDescription
	if pccRule.FlowInfos == nil {
		continue
	}
	for _, flowInfo := range *pccRule.FlowInfos {
		if flowInfo.FlowDescription == nil {
			continue
		}
		flow, err := policydb.GetFlowDescriptionFromFlowString(string(*flowInfo.FlowDescription))

```

Reply:
Done

Reply:
lets do this

```
	...
		if err != nil {
			glog.Errorf("Could not get flow for description %s : %s", *flowInfo.FlowDescription, err)
			continue
		}
		flowList = append(flowList, flow)
	}
```

The reason is you ever have to add mor stuff to that function you will have to add it inside the else. If for any reason the person that adds that new code puts it outside the else, the logic will be wrong.

Plus having a continue is very easy to tell that after that glog there is nothing else to do for that path.

Once you have change this, take a look at the function and you will see this patter of having all the edge cases at the top, and the main logic at the bottom. This way reading a function becomes very easy.